gk/fib.c

```diff
@@ -1639,20 +1664,20 @@ add_fib_entry_numerical(struct ip_prefix *prefix_info,
 		GK_LOG(NOTICE, "Prefix lookup did not find existing neighbor FIB on back interface, as expected\n");
 	}
 
-	if (gw_addr != NULL) {
+	for (i = 0; i < num_addrs; i++) {
 		/*
-		 * Verify that the IP addresses of gateways FIB entries
-		 * are not included in their prefixes.
+		 * Verify that the IP addresses of gateway FIB entries
+		 * are not included in the prefix.
 		 */
-		ret = check_gateway_prefix(prefix_info, gw_addr);
+		ret = check_gateway_prefix(prefix_info, &gw_addrs[i]);
 		if (ret < 0)
 			return -1;
 	}
```

Comment:
This pull request did not introduce either of the issues below, but they are so important that they should be addressed as an independent pull request fixing them since the first issue is a bug.
1. At this point of the code, we don't know that `prefix->addr.proto == gw_addr->proto`. But the code of `check_gateway_prefix()` is making this assumption. `check_gateway_prefix()` should always return `-1` when `prefix->addr.proto != gw_addr->proto`.
2. There is no log entry to explain what happened when `check_gateway_prefix()` fails. `check_gateway_prefix()` should log failures. The lack of a log entry explaining a failure can be very confusing in production.
3. In the case `check_gateway_prefix()` deals with an unknown protocol (i.e. neither IPv4 nor IPv6), `check_gateway_prefix()` should log and return an error instead of panicking.

We should merge the new pull request first.

Reply:
Use `unlikely()` for the expression `prefix->addr.proto != gw_addr->proto`.

Reply:
Once `check_gateway_prefix()` is enforcing `prefix->addr.proto == gw_addr->proto`, the code that does the same check in `init_gateway_fib_locked()` and `init_grantor_fib_locked()` can be replaced with a comment saying that that sanity check is already done in `add_fib_entry_numerical()` and `update_fib_entry_numerical()`.